File: src/org/warp/picalculator/math/rules/FractionsRule12.java
package org.warp.picalculator.math.rules;

import java.util.ArrayList;

import org.warp.picalculator.Error;
import org.warp.picalculator.math.functions.Division;
import org.warp.picalculator.math.functions.Expression;
import org.warp.picalculator.math.functions.Function;
import org.warp.picalculator.math.functions.Multiplication;

/**
 * Fractions rule<br>
 * <b>(b / c) / a = b / (a * c)</b>
 * 
 * @author Andrea Cavalli
 *
 */
public class FractionsRule12 {

	public static boolean compare(Function f) {
		final Division fnc = (Division) f;
		Function a;
		Function c;
		if (fnc.getVariable1() instanceof Division) {
			final Division div2 = (Division) fnc.getVariable1();
			a = fnc.getVariable1();
			c = div2.getVariable2();
			return new Multiplication(fnc.getRoot(), a, c).isSolved() == false;
		}
		return false;
	}

	public static ArrayList<Function> execute(Function f) throws Error {
		final ArrayList<Function> result = new ArrayList<>();
		final Division fnc = (Division) f;
		Function a;
		Function b;
		Function c;

		final Division div2 = (Division) fnc.getVariable1();
		a = fnc.getVariable2();
		b = div2.getVariable1();
		c = div2.getVariable2();
		result.add(new Division(fnc.getRoot(), new Multiplication(fnc.getRoot(), new Expression(fnc.getRoot(), a), new Expression(fnc.getRoot(), c)), b));

		return result;
	}

}

File: src/main/resources/rules/FractionsRule12.js
//Imports


var Error = org.warp.picalculator.Error;
var Function = org.warp.picalculator.math.Function;
var Division = org.warp.picalculator.math.functions.Division;
var Multiplication = org.warp.picalculator.math.functions.Multiplication;

var ObjectArrayList = Java.type("it.unimi.dsi.fastutil.objects.ObjectArrayList");

/**
 * Fractions rule
 * (b / c) / a = b / (a * c)
 * 
 * @author Andrea Cavalli
 *
 */
var rule = {
	// Rule name
	getRuleName: function() {
		return "FractionsRule12";
	},
	// Rule type
	getRuleType: function() {
		return RuleType.EXPANSION;
	},
	/* Rule function
	   Returns:
	     - null if it's not executable on the function "f"
	     - An ObjectArrayList<Function> if it did something
	*/

	execute: function(f) {
		var isExecutable;
		if (ScriptUtils.instanceOf(f, Division.class)) {
			var fnc = f;
			var a;
			var c;
			if (ScriptUtils.instanceOf(fnc.getParameter1(), Division.class)) {
				var div2 = fnc.getParameter1();
				a = fnc.getParameter1();
				c = div2.getParameter2();
				isExecutable = true;
			}
		}
	
		if (isExecutable) {
			var result = new ObjectArrayList();
			var fnc = f;
			var a;
			var b;
			var c;
	
			var div2 = fnc.getParameter1();
			a = fnc.getParameter2();
			b = div2.getParameter1();
			c = div2.getParameter2();
			result.add(new Division(fnc.getMathContext(), new Multiplication(fnc.getMathContext(), a, c), b));
	
			return result;
		} else {
			return null;
		}
	}
}

//Add this rule to the list of rules
RulesManager.addRule(engine.getInterface(rule, Rule.class));